Rubberduck.VBEditor.VBA/SafeComWrappers/VB/VBProject.cs

```diff
@@ -186,32 +185,116 @@ public string ProjectDisplayName
                     return _displayName;
                 }
 
-                using (var vbe = VBE)
-                using (var activeProject = vbe.ActiveVBProject)
-                using (var mainWindow = vbe.MainWindow)
+                _displayName = DisplayNameFromFileName();
+
+                if (string.IsNullOrEmpty(_displayName))
+                {
+                    _displayName = DisplayNameFromWindowCaption();
+                }
+
+                if (string.IsNullOrEmpty(_displayName)
+                    || _displayName.EndsWith("..."))
                 {
-                    try
+                    var nameFromBuildFileName = DisplayNameFromBuildFileName();
+                    if (!string.IsNullOrEmpty(nameFromBuildFileName) 
+                        && nameFromBuildFileName.Length > _displayName.Length - 3) //Otherwise, we got more of the name from the previous attempt.
                     {
-                        if (Target.HelpFile != activeProject.HelpFile)
-                        {
-                            vbe.ActiveVBProject = this;
-                        }
+                        _displayName = nameFromBuildFileName;
+                    }
+                }
+
+                return _displayName;
+            }
+        }
+
+        private string DisplayNameFromFileName()
+        {
+            return Path.GetFileName(FileName);
+        }
+
+        private string DisplayNameFromBuildFileName()
+        {
+            var pseudoDllName = Path.GetFileName(BuildFileName);
+            return pseudoDllName == null || pseudoDllName.Length <= 4 //Should not happen as the string should always end in .DLL.
+                ? string.Empty
+                : pseudoDllName.Substring(0, pseudoDllName.Length - 4);
+        }
+
+        private static readonly Regex CaptionProjectRegex = new Regex(@"^(?:[^-]+)(?:\s-\s)(?<project>.+)(?:\s-\s.*)?$");
+        private static readonly Regex OpenModuleRegex = new Regex(@"^(?<project>.+)(?<module>\s-\s\[.*\((Code|UserForm)\)\])$");
+        private static readonly Regex PartialOpenModuleRegex = new Regex(@"^(?<project>.+)(\s-\s\[)");
+        private static readonly Regex NearlyOnlyProject = new Regex(@"^(?<project>.+)(\s-?\s?)$");
+
+        private string DisplayNameFromWindowCaption()
+        {
+            using (var vbe = VBE)
+            using (var activeProject = vbe.ActiveVBProject)
+            using (var mainWindow = vbe.MainWindow)
+            {
+                try
+                {
+                    if (ProjectId != activeProject.ProjectId)
+                    {
+                        vbe.ActiveVBProject = this;
+                    }
+
+                    var caption = mainWindow.Caption;
+                    if (caption.Length > 99)
```

Comment:
I think we should consider the fact that apparently we get more caption from Win32 API than we would from `mainWindow.Caption`. Using Spy++, I was able to get up to 159 characters where `mainWindow.Caption` only gets us 98 characters. To use Win32 API, we'd need to p/invoke [`GetWindowText`](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getwindowtexta).

Reply:
I kind of don't want to do that for something that is only a fallback anyway, which again has another fallback.

Reply:
Given that this is a fallback, I think truncation to 99 characters is reasonable. You'd have to have the CE crazy-wide to even see the difference...